File: USER_SERVICE/src/main/java/com/parkito/learnmicro/monolith/service/UserService.java
package com.parkito.learnmicro.monolith.service;

import com.parkito.learnmicro.monolith.dto.UserDTO;
import com.parkito.learnmicro.monolith.entity.User;
import com.parkito.learnmicro.monolith.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.stream.Collectors;

/**
 * @author Artem Karnov @date 11/6/2017.
 * artem.karnov@t-systems.com
 */
@Service
public class UserService {
    private final UserRepository userRepository;

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }


    public boolean deleteUserByEmail(String email) {
        User userForDelete = userRepository.findByEmail(email);
        if (userForDelete != null) {
            userRepository.delete(userForDelete);
            return true;
        } else {
            return false;
        }
    }

    public UserDTO createUser(String email, String firstName, String secondName) {
        User existedUser = userRepository.findByEmail(email);
        if (existedUser == null) {
            User user = User.builder()
                    .email(email)
                    .firstName(firstName)
                    .lastName(secondName)
                    .build();
            return convert(userRepository.save(user));
        } else {
            return null;
        }
    }

    public UserDTO findUserByEmail(String email) {
        User user = userRepository.findByEmail(email);
        if (user == null) {
            return null;
        } else {
            return convert(user);
        }
    }

    private UserDTO convert(User user) {
        if (user == null) {
            return null;
        } else {
            return UserDTO.builder()
                    .email(user.getEmail())
                    .firstName(user.getFirstName())
                    .firstName(user.getLastName())
                    .build();
        }
    }
}

File: USER_SERVICE/src/main/kotlin/com/parkito/learnmicro/user/service/service/UserService.kt
package com.parkito.learnmicro.user.service.service;

import com.parkito.learnmicro.user.service.controller.RestUserClient
import com.parkito.learnmicro.user.service.dto.UserDTO
import com.parkito.learnmicro.user.service.entity.User
import com.parkito.learnmicro.user.service.exception.value.BusinessLogicException
import com.parkito.learnmicro.user.service.exception.value.ResourceNotFoundException
import com.parkito.learnmicro.user.service.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

/**
 * @author Artem Karnov @date 11/6/2017.
 * artem.karnov@t-systems.com
 */
@Service
class UserService {
    @Autowired
    lateinit var userRepository: UserRepository
    @Autowired
    lateinit var restUserClient: RestUserClient

    @Transactional
    fun createUser(email: String, firstName: String, secondName: String): UserDTO {
        val existedUser: User? = userRepository.findByEmail(email);
        if (existedUser == null) {
            val user = User(email = email, firstName = firstName, lastName = secondName)
            userRepository.save(user)
            return UserDTO.fromEntity(user, Collections.emptyList())
        } else {
            throw BusinessLogicException("User with email = $email already exists")
        }
    }

    fun findUserByEmail(email: String): UserDTO {
        val user: User? = userRepository.findByEmail(email)
        if (user != null) {
            val docs = restUserClient.getAllClientDocuments(email)
            var serials = Collections.emptyList<String>()
            if (docs != null) {
                serials = docs.map { d -> d.serial }
            }
            return UserDTO.fromEntity(user, serials)
        } else {
            throw ResourceNotFoundException("User with email = $email wasn't found")
        }
    }

    @Transactional
    fun deleteUserByEmail(email: String) {
        userRepository.deleteByEmail(email)
    }

    fun getAllUsers(): List<UserDTO> {
        return userRepository.findAll().map { u ->
            UserDTO.fromEntity(u, restUserClient.getAllClientDocuments(u.email)!!.map { s -> s.serial })
        }
    }
}
